aws/resource_aws_servicecatalog_portfolio.go

```diff
@@ -0,0 +1,145 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/servicecatalog"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsServiceCatalogPortfolio() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsServiceCatalogPortfolioCreate,
+		Read:   resourceAwsServiceCatalogPortfolioRead,
+		Update: resourceAwsServiceCatalogPortfolioUpdate,
+		Delete: resourceAwsServiceCatalogPortfolioDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Type:     schema.TypeString,
+				Computed: true,
+				ForceNew: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: false,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: false,
+			},
+			"provider_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: false,
+			},
+		},
+	}
+}
+
+func resourceAwsServiceCatalogPortfolioCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).scconn
+	input := servicecatalog.CreatePortfolioInput{}
+	if v, ok := d.GetOk("name"); ok {
+		now := time.Now()
+		input.IdempotencyToken = aws.String(fmt.Sprintf("%d", now.UnixNano()))
+		input.DisplayName = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		input.Description = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("provider_name"); ok {
+		input.ProviderName = aws.String(v.(string))
+	}
+
+	log.Printf("[DEBUG] Creating Service Catalog Portfolio: %#v", input)
+	resp, err := conn.CreatePortfolio(&input)
+	if err != nil {
+		return fmt.Errorf("Creating ServiceCatalog portfolio failed: %s", err.Error())
+	}
+	d.SetId(*resp.PortfolioDetail.Id)
+
+	return resourceAwsServiceCatalogPortfolioRead(d, meta)
+}
+
+func resourceAwsServiceCatalogPortfolioRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).scconn
+	input := servicecatalog.DescribePortfolioInput{}
+	input.Id = aws.String(d.Id())
+
+	log.Printf("[DEBUG] Reading Service Catalog Portfolio: %#v", input)
+	resp, err := conn.DescribePortfolio(&input)
+	if err != nil {
+		if scErr, ok := err.(awserr.Error); ok && scErr.Code() == "ResourceNotFoundException" {
```

Comment:
@Ninir can you suggest, or point me to a similar example, how to idiomatically write a test for this type of error in the tf test framework?

Reply:
This kind of test is not simple to implement, due to the way to test: we need to create the resource, add a custom step to remove the object, and expect a non-empty plan.

In [this example](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_customer_gateway_test.go#L50-L93), you first have a test that expects an error, and then a test that expects a non empty plan.
Depending on our implementation, we will go for one or the other (I think it will be the non-empty plan)

The best way to proceed is perhaps to `make build` the provider and check it manually.

Reply:
@ninir Ran it manually, removed the resource and it works.

```
2017/09/19 06:17:46 [WARN] Service Catalog "port-123456781234" not found, removing from state
```

Still a bit hazy but will try to construct a non-empty plan test using the above.

Reply:
@ninir added a first pass at a test.  Can you provide feedback?

https://github.com/terraform-providers/terraform-provider-aws/pull/1694/commits/037d11a42117a7e6c242a94872a241a6ce24417f

Thanks